storage/innobase/handler/handler0alter.cc

```diff
@@ -5072,6 +5081,16 @@ prepare_inplace_alter_table_dict(
 	}
 
 	if (ctx->need_rebuild() && ctx->new_table->supports_instant()) {
+		// Add versioning is:
+		// 1. add sys_trx_start, sys_trx_end columns
+		// 2. set vers_row_start, vers_row_end for dict_table_t
+		// 3. for every row set sys_trx_start = CUR_TRX_ID,
+		// sys_trx_end = MAX_TRX_ID
+		// It's not supported for instand ADD COLUMN yet.
+		if (ha_alter_info->create_info->vers_info
+			.with_system_versioning)
+			goto not_instant_add_column;
+
 		if (~ha_alter_info->handler_flags
 		    & Alter_inplace_info::ADD_STORED_BASE_COLUMN) {
 			goto not_instant_add_column;
```

Comment:
Perhaps the simplest fix is to modify `ha_innobase::check_if_supported_inplace_alter()` so that `ALGORITHM=INPLACE` will refuse to add or remove `SYSTEM VERSIONING` when `dict_index_is_auto_gen_clust(m_prebuilt->table->indexes.start)` holds. `ALGORITHM=COPY` would use the `CREATE TABLE` code path for creating the metadata.

This should not affect many users, because it is a bad idea to let InnoDB generate a hidden primary key.

Reply:
When we making existing table system-versioned table we need to set `sys_trx_start` and `sys_trx_end` for every row. As far as I understand instant ADD COLUMN if we'll walk every row this won't be instant anymore. I've tried to set default values for sys_trx_* fields but failed to make it work without crashes. Also, it doesn't seem to be logically correct.

Reply:
Let us continue this discussion in https://github.com/tempesta-tech/mariadb/issues/315